TASK: Rework asset usage for current state of CR

dlubitz opened the pull request:
Implements the asset usage based on a projection of the content repository.

dlubitz committed:
TASK: Refactor asset usage for current state of CR

dlubitz committed:
FEATURE: Global AssetUsage on configured content repositories

dlubitz committed:
BUGFIX: Use event id for log  and remove unused parameter

dlubitz committed:
TASK: Remove dependency of Neos to AssetUsage package by introducing a new interface.

dlubitz committed:
Bugfix: Re add missing use statement

dlubitz committed:
TASK: Include also asset variants into asset usage calculation

skurfuerst commented on Neos.ContentRepository.Core/Classes/SharedModel/ContentRepositoryNodeInformationInterface.php:
```diff
@@ -0,0 +1,16 @@
+use Neos\ContentRepository\Core\SharedModel\Workspace\ContentStreamId;
+use Neos\ContentRepository\Core\SharedModel\Node\NodeAggregateId;
+
+interface ContentRepositoryNodeInformationInterface
```
Uhm, I woudn't create interfaces here which are not implemented by the core.

bwaidelich replied:
I wonder if we could just omit the interface for now. Seems like a premature generalization to me at this point

dlubitz replied:
I don't think so.
It's very unclean then, to use it here: https://github.com/dlubitz/neos-development-collection/blob/8d2999b73cab3abe07ca313b4616164c7b553718/Neos.Neos/Classes/Service/ImageVariantGarbageCollector.php#L75-L75

The `UsageReference` doesn't have these methods and the `AssetUsageStrategyInterface::getUsageReferences()` returns them.

dlubitz replied:
Is removed and "replaced" by `Neos\Neos\AssetUsage\Dto\AssetUsageReference`

skurfuerst commented on Neos.ESCR.AssetUsage/Classes/Command/AssetUsageCommandController.php:
```diff
@@ -44,17 +30,26 @@ public function __construct(
     {
-        $usages = $this->assetUsageRepository->findUsages(AssetUsageFilter::create());
+        $contentRepositoryIdentifier = ContentRepositoryId::fromString($contentRepositoryIdentifier);
+        $assetServiceSyncService = $this->contentRepositoryRegistry->getService(
```
assetUsageSyncService

skurfuerst commented on Neos.ESCR.AssetUsage/Classes/Command/AssetUsageCommandController.php:
```diff
@@ -44,17 +30,26 @@ public function __construct(
+    public function syncCommand(string $contentRepositoryIdentifier = 'default', bool $quiet = false): void
     {
-        $usages = $this->assetUsageRepository->findUsages(AssetUsageFilter::create());
+        $contentRepositoryIdentifier = ContentRepositoryId::fromString($contentRepositoryIdentifier);
```
$contentRepositoryId (nitpick - sry ;) )

skurfuerst commented on Neos.ESCR.AssetUsage/Classes/Command/AssetUsageCommandController.php:
```diff
@@ -44,17 +30,26 @@ public function __construct(
      * @param bool $quiet if Set, only errors will be outputted
      */
-    public function syncCommand(bool $quiet = false): void
+    public function syncCommand(string $contentRepositoryIdentifier = 'default', bool $quiet = false): void
```
in the core, I'm trying to unify CLI arguments to just `contentRepository` :)

skurfuerst commented on Neos.ESCR.AssetUsage/Classes/Dto/NodeAddress.php:
```diff
@@ -0,0 +1,50 @@
+ * @internal
+ */
+#[Flow\Proxy(false)]
+final class NodeAddress
```
Hm, can't we use the NodeAddress here? I more closely need to check the code, but I think having such a copied object is kinda dangerous (with the same name)

dlubitz replied:
Yes, there is a NodeAddress within the FrontendRouting https://github.com/dlubitz/neos-development-collection/blob/a43ac99132245143a20fae4eed50ae9b8de58b7f/Neos.Neos/Classes/FrontendRouting/NodeAddress.php#L42

But this is just similar and can't get reused. Since this `NodeAddress` is for internal use only, the naming would not bother me, but I'm open for any suggestions!?

skurfuerst commented on Neos.ESCR.AssetUsage/Classes/Service/GlobalAssetUsageService.php:
```diff
@@ -0,0 +1,79 @@
+        );
+    }
+
+    private function getContentRepositories(): array
```
this is something I am unsure about. Maybe we can configure this per-site? (I guess easier to discuss via video call etc)

dlubitz replied:
Actually we discussed a lot, but not this place.  Do we still need to discuss this?

skurfuerst reviewed (commented):
Awesome work @dlubitz ❤️ ❤️ ❤️  It's amazing how far you got with so little guidance from us!

After thinking it through some more, I'd suggest to move *all* AssetUsage functionality
into Neos\Neos\AssetUsage (new namespace); and get rid of this extra package. I cannot
think of a scenario where only this package would be used, but Neos.Neos not.

Regarding a few of the other comments - It probably makes sense to talk through them next week if that's fine for you via a slack call :) 

Do you want to continue working on this; or should somebody of us continue here?

All the best,
Sebastian

skurfuerst commented:
Reminder to self: https://github.com/neos/neos-development-collection/pull/4077 needs this one here as a basis.

dlubitz commented:
Thanks for your feedback @skurfuerst. I would like to finish that, so let have a slack call to discuss it. Would be great to have @bwaidelich with us, as he gave the main input for these changes.

skurfuerst commented:
@dlubitz awesome - so let's tomorrow coordinate over Slack and find a time slot :)

dlubitz committed:
Merge remote-tracking branch 'upstream/9.0' into feature/90/asset-usage

dlubitz committed:
TASK: Fix naming

dlubitz committed:
TASK: Move from Neos.ESCR.AssetUsage into Neos.Neos

dlubitz commented:
I moved the code into the Neos.Neos Package, but where shall we put the Readme file of this package? @skurfuerst 
https://github.com/dlubitz/neos-development-collection/blob/806d25fe355ec5426c2a9d0a963cfce7c949213c/Neos.ESCR.AssetUsage/README.md#L44-L44

bwaidelich commented:
@dlubitz Awesome, thanks!

> where shall we put the Readme file of this package?

I'd say `Neos.Neos/Classes/AssetUsage/README.md`

dlubitz committed:
TASK: Move from Neos.ESCR.AssetUsage into Neos.Neos

dlubitz committed:
TASK: Move from Neos.ESCR.AssetUsage into Neos.Neos

dlubitz committed:
FEATURE: Store original asset id in projection

dlubitz committed:
TASK: Mark classes as internal or api

bwaidelich commented on Neos.Neos/Classes/AssetUsage/Projection/AssetUsageProjection.php:
```diff
@@ -19,24 +19,50 @@
+            $dbal,
+            $this->repository->getTableNamePrefix() . '_checkpoint',
+            self::class
+        );
```
Just a thought: Instead of exposing `getTableNamePrefix()` we could let the interface implement the `CheckpointStorageInterface` (adapter for the `DoctrineCheckpointStorage` with the prefix it already knows) and then replace `$this->checkpointStorage->` by `$this->repository->` in this class.

dlubitz replied:
Not sure, what you mean here. Do you want to implement a "getDoctrineCheckpointStorage" on the Repository?

bwaidelich replied:
No, sth along the lines of
```php
final class AssetUsageRepository implements CheckpointStorageInterface
{
    private DoctrineCheckpointStorage $checkpointStorage;

    public function __construct(
        private readonly Connection $dbal,
        private readonly string $tableNamePrefix,
    ) {
        $this->checkpointStorage = new DoctrineCheckpointStorage($dbal, $tableNamePrefix . '_checkpoint', AssetUsageProjection::class);
    }

    public function acquireLock(): SequenceNumber {
        return $this->checkpointStorage->acquireLock();
    }
    public function updateAndReleaseLock(SequenceNumber $sequenceNumber): void {
        $this->checkpointStorage->updateAndReleaseLock($sequenceNumber);
    }
    public function getHighestAppliedSequenceNumber(): SequenceNumber {
        return $this->checkpointStorage->getHighestAppliedSequenceNumber();
    }

// ...
}
```

I'm not 100% sure if it's worth it and whether it is much cleaner at all, but it would at least keep the table name prefix local

dlubitz replied:
Hmm, somehow interesting to have it in the repository. But this would bind CheckpointStorage to the same storage, as the repository, which I assume is not necessary. OTOH it's bound to each other by the projection anyways. 🤷‍♂️ 

Maybe @skurfuerst has an opinion on this?

dlubitz replied:
General question: Why do we need a separate table per projection? Wouldn't is work with one table for all checkpoints?

bwaidelich replied:
> General question: Why do we need a separate table per projection?

Good point! I don't remember whether this was a requirement for the table lock => https://github.com/neos/neos-development-collection/issues/4152

bwaidelich commented on Neos.Neos/Classes/AssetUsage/Service/GlobalAssetUsageService.php:
```diff
@@ -0,0 +1,93 @@
+
+    public function findAssetUsageByAssetId(string $assetId): AssetUsages
+    {
+        $assetUsages = $this->withAllRepositories(
+            function (ContentRepository $repository) use ($assetId) {
+                $filter = AssetUsageFilter::create()
+                    ->withAsset($assetId)
+                    ->includeVariantsOfAsset()
+                    ->groupByNode();
+
+                return $repository->projectionState(AssetUsageFinder::class)->findByFilter($filter);
+            }
+        );
+
+        return AssetUsages::fromArrayOfAssetUsages($assetUsages);
```
I find this construct to make things less readable for no imminent gain (AFAICS).

Why not just:

```suggestion
        $filter = AssetUsageFilter::create()
            ->withAsset($assetId)
            ->includeVariantsOfAsset()
            ->groupByNode();
        $assetUsages = [];
        foreach ($this->getContentRepositories() as $contentRepository) {
            $assetUsages[] = $repository->projectionState(AssetUsageFinder::class)->findByFilter($filter);
        }
        return AssetUsages::fromArrayOfAssetUsages(array_merge(...$assetUsages));
```

dlubitz replied:
Sure, I'll replace it.

bwaidelich commented on Neos.Neos/Classes/AssetUsage/Service/GlobalAssetUsageService.php:
```diff
@@ -0,0 +1,93 @@
+     * @param callable $callback
+     * @return array<mixed>
+     */
+    private function withAllRepositories(callable $callback): array
```
I would suggest to remove this (see comment above)

dlubitz replied:
Done!

bwaidelich commented on Neos.Neos/Classes/AssetUsage/Service/GlobalAssetUsageService.php:
```diff
@@ -0,0 +1,93 @@
+    public function removeAssetUsageByAssetId(string $assetId): void
+    {
+        $this->withAllRepositories(
+            fn(AssetUsageRepository $repository) => $repository->removeAsset($assetId)
```
`$repository` is an instance of `ContentRepository` according to the code and there is no `ContentRepository::removeAsset()` or what did I miss?

dlubitz replied:
You are right, this is a bug. I fixed it. Just wondering, why this error didn't apper during testing.

bwaidelich commented on Neos.Neos/Classes/Service/ImageVariantGarbageCollector.php:
```diff
@@ -62,24 +62,24 @@ public function removeUnusedImageVariant(Node $node, $propertyName, $oldValue, $
-                    && $usageItem->originDimensionSpacePoint === $node->originDimensionSpacePoint->hash
-                    && $usageItem->nodeAggregateIdentifier === $node->nodeAggregateId
+                    $usageItem instanceof ContentRepositoryNodeInformationInterface
+                    && $usageItem->getContentStreamId() === $node->subgraphIdentity->contentStreamId
```
Careful, two Content Stream VOs share the same instance but we should never rely on that fact
This should either be
```suggestion
                    && $usageItem->getContentStreamId()->equals($node->subgraphIdentity->contentStreamId)
```
or
```suggestion
                    && $usageItem->getContentStreamId()->value === $node->subgraphIdentity->contentStreamId->value
```

dlubitz replied:
Was already there, but is fixed now.

bwaidelich commented on Neos.Neos/Classes/Service/ImageVariantGarbageCollector.php:
```diff
@@ -62,24 +62,24 @@ public function removeUnusedImageVariant(Node $node, $propertyName, $oldValue, $
+                    $usageItem instanceof ContentRepositoryNodeInformationInterface
+                    && $usageItem->getContentStreamId() === $node->subgraphIdentity->contentStreamId
+                    && $usageItem->getOriginDimensionSpacePointHash() === $node->originDimensionSpacePoint->hash
+                    && $usageItem->getNodeAggregateId() === $node->nodeAggregateId
```
Same here and in this case it won't even work because:
```php
NodeAggregateId::fromString('foo') !== NodeAggregateId::fromString('foo');
```
```suggestion
                    && $usageItem->getNodeAggregateId()->equals($node->nodeAggregateId)
```

bwaidelich reviewed (commented):
This already looks great, thanks a lot for all your work!
Just a few thoughts and suggestions

dlubitz committed:
BUGFIX: Comparison of ContentStreamId and NodeAggregateId with "equals"

dlubitz committed:
TASK: Remove "withAllRepositories" and replace with foreach and fix asset usage removal

bwaidelich commented on Neos.Neos/Classes/AssetUsage/Service/GlobalAssetUsageService.php:
```diff
@@ -82,12 +87,12 @@ private function getContentRepositories(): array
+    private function getAssetUsageRepository(ContentRepositoryId $contentRepositoryId): AssetUsageRepository
     {
-        return array_map(fn($repository) => $callback($repository), $this->getContentRepositories());
+        if (!array_key_exists((string)$contentRepositoryId, $this->assetUsageRepositories)) {
```
```suggestion
        if (!array_key_exists($contentRepositoryId->value, $this->assetUsageRepositories)) {
```

bwaidelich commented on Neos.Neos/Classes/AssetUsage/Service/GlobalAssetUsageService.php:
```diff
@@ -82,12 +87,12 @@ private function getContentRepositories(): array
     {
-        return array_map(fn($repository) => $callback($repository), $this->getContentRepositories());
+        if (!array_key_exists((string)$contentRepositoryId, $this->assetUsageRepositories)) {
+            $this->assetUsageRepositories[(string)$contentRepositoryId] = $this->assetUsageRepositoryFactory->build($contentRepositoryId);
```
```suggestion
            $this->assetUsageRepositories[$contentRepositoryId->value] = $this->assetUsageRepositoryFactory->build($contentRepositoryId);
```

bwaidelich commented on Neos.Neos/Classes/AssetUsage/Service/GlobalAssetUsageService.php:
```diff
@@ -82,12 +87,12 @@ private function getContentRepositories(): array
+            $this->assetUsageRepositories[(string)$contentRepositoryId] = $this->assetUsageRepositoryFactory->build($contentRepositoryId);
+        }
+
+        return $this->assetUsageRepositories[(string)$contentRepositoryId];
```
```suggestion
        return $this->assetUsageRepositories[$contentRepositoryId->value];
```

bwaidelich commented on Neos.Neos/Classes/AssetUsage/Service/GlobalAssetUsageService.php:
```diff
@@ -0,0 +1,98 @@
+                }
+                $contentRepositoryId = ContentRepositoryId::fromString($contentRepositoryId);
+
+                $this->repositories[(string)$contentRepositoryId] = $this->contentRepositoryRegistry->get(
```
```suggestion
                $this->repositories[$contentRepositoryId->value] = $this->contentRepositoryRegistry->get(
```

bwaidelich reviewed (commented):
Sorry, overlooked at first iteration: We should avoid implicit string casting and we plan to remove `__toString()` from VOs for greater type safety

dlubitz committed:
TASK: Use explicit value instead of __toString

dlubitz committed:
TASK: Fix typos

bwaidelich commented:
maybe we can incorporate this with https://github.com/neos/neos-development-collection/pull/3333 – I'll look into it asap

dlubitz committed:
TASK: Fix lint errors

dlubitz committed:
TASK: Fix lint errors

dlubitz committed:
TASK: Fix lint errors

bwaidelich committed:
Merge branch '9.0' into feature/90/asset-usage

bwaidelich committed:
Skip removed properties in AssetUsageProjection

bwaidelich commented:
## Todos as discussed with @dlubitz 

- [ ] Introduce `AssetUsagesByContentRepository` that contains the CR ids as keys
- [ ] Adjust `\Neos\Neos\AssetUsage\Service\GlobalAssetUsageService::findAssetUsageByAssetId` and `\Neos\Neos\AssetUsage\AssetUsageStrategy::getUsages` to use new type as return type
- [ ] Remove `\Neos\Neos\Domain\Model\Dto\AssetUsageInNodeProperties` (replace usages with `\Neos\Neos\AssetUsage\Dto\AssetUsageReference`)
- [ ] Remove `\Neos\ContentRepository\Core\SharedModel\ContentRepositoryNodeInformationInterface`
- [ ] Adjust `\Neos\Media\Browser\Controller\UsageController` accordingly

skurfuerst commented:
@dlubitz fixed "my" issue with the content streams - no further action from you should be necessary :)

dlubitz committed:
Merge remote-tracking branch 'upstream/9.0' into feature/90/asset-usage

dlubitz committed:
TASK: Remove ContentRepositoryNodeInformationInterface, store OriginDimensionSpacePoint in projection

dlubitz committed:
Merge remote-tracking branch 'upstream/9.0' into feature/90/asset-usage

dlubitz committed:
TASK: Cleanup

dlubitz committed:
BUGFIX: Return usage references flat instead of grouped by content repository

dlubitz committed:
TASK: Adpot asset usage backend view to new CR

dlubitz commented:
All todos (https://github.com/neos/neos-development-collection/pull/4088#issuecomment-1491890252) are done.

bwaidelich committed:
Merge branch '9.0' into feature/90/asset-usage

bwaidelich committed:
Add `originalassetid` to unique index of asset_usage table

..to work around a [bug](https://jira.mariadb.org/browse/MDEV-22756) in MariaDB.

bwaidelich committed:
Make copyContentStream() and copyDimensions() independant from column order

Adjusts `\Neos\Neos\AssetUsage\Projection\AssetUsageRepository::copyContentStream()`
and `\Neos\Neos\AssetUsage\Projection\AssetUsageRepository::copyDimensions()` such
that they work independantly from the column order

bwaidelich committed:
Mostly cosmetic tweaks and cleanups

bwaidelich committed:
Cosmetic fix to satisfy linter

bwaidelich reviewed (approved):
Awesome work!

bwaidelich reviewed (approved):
Awesome work, thanks @dlubitz !
